test/cypress/integration/card/token-reuse.spec.js

```diff
@@ -0,0 +1,164 @@
+const cardPaymentStubs = require('../../utils/card-payment-stubs')
+
+describe('Re-use token flow', () => {
+  const tokenId = 'be88a908-3b99-4254-9807-c855d53f6b2b'
+  const chargeId = 'ub8de8r5mh4pb49rgm1ismaqfv'
+  const createPaymentChargeStubs = cardPaymentStubs.buildCreatePaymentChargeStubs(tokenId, chargeId)
+  const usedTokenAndReturnPaymentCreatedChargeStubs = cardPaymentStubs.buildUsedTokenAndReturnPaymentChargeStubs(tokenId, chargeId, 'ENTERING CARD DETAILS')
+  const usedTokenAndReturnPaymentAuthSuccessChargeStubs = cardPaymentStubs.buildUsedTokenAndReturnPaymentChargeStubs(tokenId, chargeId, 'AUTHORISATION SUCCESS')
+  const usedTokenAndReturnPaymentExpiredChargeStubs = cardPaymentStubs.buildUsedTokenAndReturnPaymentChargeStubs(tokenId, chargeId, 'EXPIRED')
+
+  const checkCardDetailsStubs = [
+    { name: 'connectorGetChargeDetails',
+      opts: {
+        chargeId,
+        status: 'ENTERING CARD DETAILS',
+        state: { finished: false, status: 'started' }
+      }
+    },
+    { name: 'cardIdValidCardDetails' }
+  ]
+
+  const validPayment = {
+    cardNumber: '4444333322221111',
+    expiryMonth: '01',
+    expiryYear: '30',
+    name: 'Valid Paying Name',
+    securityCode: '012',
+    addressLine1: '10 Valid Paying Address',
+    city: 'London',
+    postcode: 'E1 8QS',
+    email: 'validpayingemail@example.com'
+  }
+
+  const confirmPaymentDetailsStubs = [
+    { name: 'adminUsersGetService', opts: {} },
+    { name: 'connectorMultipleSubsequentChargeDetails',
+      opts: [{
+        chargeId,
+        status: 'ENTERING CARD DETAILS',
+        state: { finished: false, status: 'started' }
+      }, {
+        chargeId,
+        paymentDetails: validPayment,
+        status: 'AUTHORISATION SUCCESS',
+        state: { finished: false, status: 'submitted' }
+      }]
+    },
+    { name: 'cardIdValidCardDetails' },
+    { name: 'connectorValidPatchConfirmedChargeDetails', opts: { chargeId } }
+  ]
+
+  beforeEach(() => {
+    // this test is for the full process, the session should be maintained
+    // as it would for an actual payment flow
+    Cypress.Cookies.preserveOnce('frontend_state')
+  })
+
+  describe('Visiting the secure url for the first time', () => {
+    it('should load the Enter Card Details page', () => {
+      cy.task('setupStubs', createPaymentChargeStubs)
+      cy.visit(`/secure/${tokenId}`)
+
+      // 1. Charge will be created using this id as a token (GET)
+      // 2. Token will be marked as used (POST)
+      // 3. Charge will be fetched (GET)
+      // 4. Service related to charge will be fetched (GET)
+      // 5. Charge status will be updated (PUT)
+      // 6. Client will be redirected to /card_details/:chargeId (304)
+      cy.location('pathname').should('eq', `/card_details/${chargeId}`)
+    })
+  })
+
+  describe('Visiting the secure url again', () => {
```

Comment:
Might be clearer if these mentioned the specific states e.g. “Visiting the secure URL again when the state is ENTERING CARD DETAILS”.

Reply:
i think this is a good idea and i've updated the test names with the states